libyara/modules/dex.c

```diff
@@ -0,0 +1,650 @@
+    declare_integer("superclass_idx");
+    declare_integer("interfaces_offset");
+    declare_integer("source_file_idx");
+    declare_integer("annotations_offset");
+    declare_integer("class_data_offset");
+    declare_integer("static_values_offset");
+  end_struct_array("class_defs");
+
+  begin_struct_array("proto_ids");
+    declare_integer("shorty_idx");
+    declare_integer("return_type_idx");
+    declare_integer("parameters_offset");
+  end_struct_array("proto_ids");
+
+  begin_struct_array("field_ids");
+    declare_integer("class_idx");
+    declare_integer("type_idx");
+    declare_integer("name_idx");
+  end_struct_array("field_ids");
+
+  begin_struct_array("method_ids");
+    declare_integer("class_idx");
+    declare_integer("proto_idx");
+    declare_integer("name_idx");
+  end_struct_array("method_ids");
+
+  begin_struct("map_list");
+    declare_integer("size");
+    begin_struct_array("map_items");
+      declare_integer("type");
+      declare_integer("unused");
+      declare_integer("size");
+      declare_integer("offset");
+    end_struct_array("map_items");
+  end_struct("map_list");
+
+end_declarations;
+
+PDEX_HEADER dex_get_header(uint8_t *data, size_t data_size);
+void load_header(PDEX_HEADER dex_header, YR_OBJECT *module);
+void load_string_ids(PDEX_HEADER dex_header, uint8_t *data, size_t data_size, YR_OBJECT *module);
+void load_type_ids(PDEX_HEADER dex_header, uint8_t *data, size_t data_size, YR_OBJECT *module);
+void load_class_defs(PDEX_HEADER dex_header, uint8_t *data, size_t data_size, YR_OBJECT *module);
+void load_proto_ids(PDEX_HEADER dex_header, uint8_t *data, size_t data_size, YR_OBJECT *module);
+void load_field_ids(PDEX_HEADER dex_header, uint8_t *data, size_t data_size, YR_OBJECT *module);
+void load_method_ids(PDEX_HEADER dex_header, uint8_t *data, size_t data_size, YR_OBJECT *module);
+void load_map_list(PDEX_HEADER dex_header, uint8_t *data, size_t data_size, YR_OBJECT *module);
+char *get_string_item(uint32_t index, YR_OBJECT *module);
+char *get_prototype_string(uint16_t proto_idx, uint8_t *data, size_t data_size, YR_OBJECT *module);
+uint64_t read_uleb128(uint8_t **buf, unsigned *uleb_size);
+uint64_t get_bytes_until_null(const uint8_t **buf);
+void print_hex_arr(uint8_t *buf, int len);
+bool is_dex_corrupt(PDEX_HEADER dex_header);
+
+int module_initialize(
+    YR_MODULE *module)
+{
+  return ERROR_SUCCESS;
+}
+
+
+int module_finalize(
+    YR_MODULE *module)
+{
+  return ERROR_SUCCESS;
+}
+
+
+int module_load(
+    YR_SCAN_CONTEXT* context,
+    YR_OBJECT* module_object,
+    void* module_data,
+    size_t module_data_size)
+{
+  YR_MEMORY_BLOCK* block;
+  YR_MEMORY_BLOCK_ITERATOR* iterator = context->iterator;
+
+  foreach_memory_block(iterator, block)
+  {
+    uint8_t* block_data = block->fetch_data(block);
+    if (block_data == NULL) {
+      continue;
+    }
+
+    PDEX_HEADER dex_header = dex_get_header(block_data, block->size);
+    if (dex_header != NULL) {
```

Comment:
It looks like pe.c and dotnet.c will fail gracefully by returning SUCCESS if the header's can't be parsed. I'm just trying to simulate this. If I don't do this and return INVALID_FILE or something, any rule which imports dex and doesn't explicitly check the dex magic itself will error out. Am I misreading how pe.c and other modules are supposed to work?